text/0021-any-and-all-operators.md

```diff
@@ -18,14 +18,14 @@ This RFC proposes extending the language with `all` and `any` operators that all
 
 ## Basic example
 
-Consider writing a policy that allows access when all port numbers in the `context` are between 8000 and 8999.  This can't be expressed in Cedar today.
+Consider writing a policy that allows access when all port numbers in the `context` are greater than or equal to 8000.  This can't be expressed in Cedar today.
 
 With the `all` operator, we can write this policy as follows:
 
 ```
 permit(principal, action, resource)
 when {
-  context.portNumbers.all(it >= 8000 && it <= 8999)
+  context.portNumbers [all >= 8000]
 };
 ```
```

Comment:
I'm a little sad that in this revision, we lost the convenient/nice syntax for examples like number-in-range.  Instead of a single `all` operation expressing that all the numbers are in a given range, we have to use two separate `all` operations, as discussed farther below in this RFC.  I realize that no expressivity is gained just by adding back `&&` and `||`, but I wonder if it would still be more ergonomic to support `&&` and `||` for use cases such as this?  Under the hood, we could even just desugar it into multiple `any` or `all` -- for instance,
```
context.portNumbers [all (it >= 8000 && it <= 8999)]
```
could literally desugar into
```
context.portNumbers [all >= 8000] &&
context.portNumbers [all <= 8999]
```
Thoughts on this?

Reply:
I prefer removing `it` - it's an implicit-context keyword (e.g. if you know what "it" is from other languages, you grok it, but if you don't then it's another thing to explain). I think removing `it` and allowing for composition with `&&` and `||` allows for a more consistent overall language and reduced overall conceptual overhead.

Reply:
Yeah, I'm a bit sad about this one too.  

Another idea (and the syntax of this needs finessing) is to add an operator to the language that lets you express range constraints nicely.  @mwhicks1 says Ruby has an operator that does this nicely; we could take inspiration from their design.  This operator would be useful for Cedar anyway, so that we can write something like `context.number inRange [1..10]` versus having to write `context.number >= 1 && context.number <= 10`.

Reply:
Ruby has a "range" object: https://ruby-doc.org/core-2.5.1/Range.html. Instead of literally making such an object we could just steal its syntax as Emina has proposed, e.g., `context.portNum inRange (1..10)` or `context.portNum inRange (1...10)`, where `..` is "inclusive" (1 up to and including 10) and `...` is exclusive (1 up to but excluding 10).

Reply:
I will say that Ruby's range operators are an incessant source of bugs and confusion—is it an inclusive or exclusive range? Is it a Range object or translated to an Array immediately? I've written so many bugs because of the syntax 😞  I found [this listing of various range syntaxes](https://rigaux.org/language-study/syntax-across-languages.html#VrsDatTpsRng) and honestly none of them are great though, so if Cedar does use a Range syntax (and I agree it's helpful), I'd argue for using either inclusive or exclusive but not both.

Stepping back from numeric ranges though, I imagine there there are other any/all set membership conditions that might want to be expressed that aren't numeric. I don't think `it` syntax really solves for those (and it also implies that Cedar would allow for variable binding..), but continued application of `&&`, `||` , and `!` certainly will. I'd prefer to avoid the pitfalls of ranges and also keep the doors open to more complex sets by relying entirely on setting ranges with logical operators. I also don't believe they're any harder to grok. IMO `element.all? >= 4 && element.all? < 8` is much clearer than `element.all? inRange [4..8]` (or is that `[4..9]`? or `[3..8]`? 😅)